### PR TITLE
Reuse StringLatin1::equals in regionMatches

### PR DESCRIPTION
**Pull request by franz1981**
This improvement has been found on https://github.com/vert-x3/vertx-web/pull/2526.

It can potentially affect the existing ArraysSupport.mismatch caller code-path performance ie requires investigation.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16933/head:pull/16933` \
`$ git checkout pull/16933`

Update a local copy of the PR: \
`$ git checkout pull/16933` \
`$ git pull https://git.openjdk.org/jdk.git pull/16933/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16933`

View PR using the GUI difftool: \
`$ git pr show -t 16933`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16933.diff">https://git.openjdk.org/jdk/pull/16933.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -2177,10 +2177,14 @@ public boolean regionMatches(int toffset, String other, int ooffset, int len) {
         if (len <= 0) {
            return true;
         }
+        byte coder = coder();
+        byte otherCoder = other.coder();
         byte[] tv = value;
         byte[] ov = other.value;
-        byte coder = coder();
-        if (coder == other.coder()) {
+        if (coder == otherCoder) {
```
Comment (schlosna) [inline src/java.base/share/classes/java/lang/String.java:2184]: Is this change needed as `otherCoder` is not reused?

```suggestion
        byte[] tv = value;
        byte[] ov = other.value;
        byte coder = coder();
        if (coder == other.coder()) {
```

File: src/java.base/share/classes/java/lang/String.java [hunk 2]
```diff
@@ -2177,10 +2177,14 @@ public boolean regionMatches(int toffset, String other, int ooffset, int len) {
...
+            if (ooffset == 0 && toffset == 0 && len == (tv.length >> coder) && ov.length == tv.length) {
+                return StringLatin1.equals(tv, ov);
```
[the remark this answers is not in the document]
Comment (franz1981) [reply src/java.base/share/classes/java/lang/String.java:2186]: > Could we save a comparison by ORing offsets?

I'm ok to use the or to save an addition offset comparison, good idea!

> Should this also handle UTF-16?

Currently String::equals doesn't use the UTF16 variant of the intrinsic, see https://github.com/openjdk/jdk/blob/949846986f572dfb82912e7d71e7bfd37a90871e/src/java.base/share/classes/java/lang/String.java#L1879-L1880; I think it's because checking byte per byte equality (from a semantic perspective), still hold. I didn't checked what are the other usage for the UTF16 version, but @cl4es could clarify it .

> The equals method will do the tv and ov length comparisons themselves, could we remove that check here or is the method call overhead too large?

I didn't remove it because region equality won't return (always) false, as equals would do, in such case, but will compare the (If existing) minimal common available length between the two.

> Do you happen to have results for make test TEST="micro:java.lang.StringComparisons"?

Not yet, I have performed a microbenchmark myself to verify that with small strings (below 16 bytes), equals was much better (both by enabling and disabling the mismatch intrinsics), but I will do it, thanks!

I'm missing as well to verify the bytecode size of the method after my changes, to make sure I didn't make it non-inlineable.

File: src/java.base/share/classes/java/lang/String.java (hunk 1 above)
Comment (franz1981) [reply src/java.base/share/classes/java/lang/String.java:2184]: Not necessary, but given that later, in the path I have introduced, I am using directly the coder byte to both compute the actual length of `tv` (inlining String::length) and verify the coder to be the same for both `tv` and `ov`, I have avoided fetching it twice.
I see that "could" increase the register pressure, but I haven't verified it yet; there is any performance concern or is a readability/style-only concern?

File: src/java.base/share/classes/java/lang/String.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cl4es) [reply src/java.base/share/classes/java/lang/String.java:2186]: Yes, since the `coder` is checked first we know the bytes compared are either both latin1 or both UTF16 so there's no need for a UTF-16-aware char-by-char comparison here. In fact the `StringUTF16.equals` method looks completely unused.

[the remark this answers is not in the document]
Comment (franz1981) [reply src/java.base/share/classes/java/lang/String.java:2186]: I can send a separate PR to remove it in case, unless we have any plans to use it elsewhere

**Comment by franz1981** [comment]
@schlosna 

Running `TEST="micro:java.lang.StringComparisons.regionMatches"` on AMD 7950x@4.5 GHz with tuned network-latency profile on and turbo-boost disabled using `numactl --localalloc -N 0` to avoid weird NUMA-like effects on heap objects.

baseline at 25f9af99be1c906fc85b8192df8fa50cced3474f:
```
Benchmark                             (size)  (utf16)  Mode  Cnt    Score   Error  Units
StringComparisons.regionMatches            6     true  avgt    5    4.380 ? 0.030  ns/op
StringComparisons.regionMatches            6    false  avgt    5    5.772 ? 0.056  ns/op
StringComparisons.regionMatches           15     true  avgt    5    4.005 ? 0.104  ns/op
StringComparisons.regionMatches           15    false  avgt    5    4.030 ? 0.055  ns/op
StringComparisons.regionMatches         1024     true  avgt    5   30.037 ? 0.089  ns/op
StringComparisons.regionMatches         1024    false  avgt    5   17.734 ? 0.092  ns/op
StringComparisons.regionMatchesRange       6     true  avgt    5    4.825 ? 0.067  ns/op
StringComparisons.regionMatchesRange       6    false  avgt    5    5.878 ? 0.056  ns/op
StringComparisons.regionMatchesRange      15     true  avgt    5    5.736 ? 0.069  ns/op
StringComparisons.regionMatchesRange      15    false  avgt    5    5.447 ? 0.028  ns/op
StringComparisons.regionMatchesRange    1024     true  avgt    5   31.169 ? 0.009  ns/op
StringComparisons.regionMatchesRange    1024    false  avgt    5   16.614 ? 0.168  ns/op
```
With 1bd619a5bd2faa8057cb85105b2c9b4997fbf2ac:
```
Benchmark                             (size)  (utf16)  Mode  Cnt    Score   Error  Units
StringComparisons.regionMatches            6     true  avgt    5    3.535 ? 0.022  ns/op
StringComparisons.regionMatches            6    false  avgt    5    3.134 ? 0.022  ns/op
StringComparisons.regionMatches           15     true  avgt    5    2.568 ? 0.022  ns/op
StringComparisons.regionMatches           15    false  avgt    5    3.415 ? 0.017  ns/op
StringComparisons.regionMatches         1024     true  avgt    5   30.052 ? 0.070  ns/op
StringComparisons.regionMatches         1024    false  avgt    5   17.024 ? 0.111  ns/op
StringComparisons.regionMatchesRange       6     true  avgt    5    4.819 ? 0.010  ns/op
StringComparisons.regionMatchesRange       6    false  avgt    5    5.888 ? 0.083  ns/op
StringComparisons.regionMatchesRange      15     true  avgt    5    5.849 ? 0.106  ns/op
StringComparisons.regionMatchesRange      15    false  avgt    5    5.466 ? 0.069  ns/op
StringComparisons.regionMatchesRange    1024     true  avgt    5   31.177 ? 0.015  ns/op
StringComparisons.regionMatchesRange    1024    false  avgt    5   16.872 ? 0.387  ns/op
```
Which translate in a ~1.8 speedup for small sized ones (which is still a fairly common use case), while bigger ones seems unchanged. 
This can be either due to caching effects becoming more relevant and/or a better behaviour of the equals intrinsic
for tail/remaining bytes, if compared to `mismatch`: I need to `perfasm` to check what's going on.

File: src/java.base/share/classes/java/lang/String.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cl4es) [reply src/java.base/share/classes/java/lang/String.java]: `StringUTF16::equals` was used before https://bugs.openjdk.org/browse/JDK-8215017 - I don't remember exactly what performance verification I did back then. x86 intrinsics seem to emit the exact same asm, aarch64 does a little bit of extra optimization for the UTF16 case (avoids code and a branch for checking tailing byte, but only if the UTF16 string is 1-3 elements long). Perhaps we should double-check that JDK-8215017 didn't significantly regress some UTF16 equality checks before we rip out code.

**Comment by franz1981** [comment]
The new commit, introducing the full positive use case (maybe relevant for the case with few characters), and adding
and additional comparison vs the `String::equals` case.

baseline at 25f9af99be1c906fc85b8192df8fa50cced3474f:
```
Benchmark                            (size)  (utf16)  Mode  Cnt   Score   Error  Units
StringComparisons.same                    6     true  avgt    5   2.402 ? 0.028  ns/op
StringComparisons.same                    6    false  avgt    5   2.056 ? 0.056  ns/op
StringComparisons.same                   15     true  avgt    5   3.733 ? 0.161  ns/op
StringComparisons.same                   15    false  avgt    5   2.807 ? 0.214  ns/op
StringComparisons.same                 1024     true  avgt    5  23.485 ? 0.150  ns/op
StringComparisons.same                 1024    false  avgt    5  15.302 ? 0.232  ns/op

StringComparisons.sameRegionMatches       6     true  avgt    5   4.410 ? 0.078  ns/op
StringComparisons.sameRegionMatches       6    false  avgt    5   5.414 ? 0.028  ns/op
StringComparisons.sameRegionMatches      15     true  avgt    5   5.770 ? 0.021  ns/op
StringComparisons.sameRegionMatches      15    false  avgt    5   5.771 ? 0.035  ns/op
StringComparisons.sameRegionMatches    1024     true  avgt    5  30.964 ? 0.023  ns/op
StringComparisons.sameRegionMatches    1024    false  avgt    5  16.807 ? 0.181  ns/op
```
with 1bd619a5bd2faa8057cb85105b2c9b4997fbf2ac:
```
Benchmark                            (size)  (utf16)  Mode  Cnt   Score   Error  Units
StringComparisons.sameRegionMatches       6     true  avgt    5   3.442 ? 0.016  ns/op
StringComparisons.sameRegionMatches       6    false  avgt    5   3.117 ? 0.002  ns/op
StringComparisons.sameRegionMatches      15     true  avgt    5   4.759 ? 0.075  ns/op
StringComparisons.sameRegionMatches      15    false  avgt    5   3.813 ? 0.026  ns/op
StringComparisons.sameRegionMatches    1024     true  avgt    5  28.308 ? 0.058  ns/op
StringComparisons.sameRegionMatches    1024    false  avgt    5  16.774 ? 0.220  ns/op
```
As confirmed by the previous results, the better value of this PR is with small-sized strings, but it is yet to be verified if:
- due to a better tail-handing of the `equals` intrinsics
- larger strings are limited by the amount of cache activity

Running `perfnorm` against 25f9af99be1c906fc85b8192df8fa50cced3474f with `same` vs `sameRegionMatches` reveal that 
both have high IPC with an high number of branche AND L1 cache-loads (with nearly no misses), which means that the amount of branches and cache-loads dominate the cost of computation despite the number of instructions on `sameRegionMatches` is higher.
In short, with bigger sized-strings, the difference between the 2 instrincs just fade away.

Now forcing some tail execution on mid-long sized ones to verify if there's some better handling of tail bytes from by `same`.

File: src/java.base/share/classes/java/lang/String.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cl4es) [reply src/java.base/share/classes/java/lang/String.java]: Going out on this tangent... I've compared the mainline (Base) with a build where `String::equals` is restored to pre-JDK-8215017 using a modified version of the StringEquals microbenchmark that tests UTF16 Strings of size 3. One test (EQ) where the strings are equals, another (NE) where they are not. The EQ one is the main contender for a case that would benefit from avoiding the trailing byte check on aarch64:

On my M1 (aarch64):

```Name                          Cnt  Base   Error   Test   Error  Unit  Change
StringEquals.equalsUTF16_3_EQ   5 1,750 ± 0,009  1,864 ± 0,401 ns/op   0,94x (p = 0,070 )
StringEquals.equalsUTF16_3_NE   5 1,674 ± 0,026  1,839 ± 0,179 ns/op   0,91x (p = 0,001*)
  * = significant
 ```

Similarly restoring the pre-JDK-8215017 version seem to be a net loss on x86:
```
Name                          Cnt  Base   Error   Test   Error  Unit  Change
StringEquals.equalsUTF16_3_EQ   5 2.885 ± 0.054  2.886 ± 0.036 ns/op   1.00x (p = 0.837 )
StringEquals.equalsUTF16_3_NE   5 2.581 ± 0.002  2.756 ± 0.003 ns/op   0.94x (p = 0.000*)
  * = significant
```

So it seems JDK-8215017 was either neutral or a small performance win (phew!). The avoided branch and overall reduction in code complexity outweighs the win (if any) from not having the redundant `COMPARE_BYTE` chunk of code emitted in the StringUTF16 case. There are other platforms I can't currently check, but I think it's likely that we'd see similar numbers there. So it seems to be the case that `StringUTF16::equals` can be safely removed.

**Comment by franz1981** [comment]
@cl4es I can undraft this, but I have no powers to create a JDK issue myself :"/

File: src/java.base/share/classes/java/lang/String.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (franz1981) [reply src/java.base/share/classes/java/lang/String.java]: Nice one, I can happily send a PR for it, but I feel bad about it; you have performed the check yourself (and were aware of what dropping it would mean...)

[the remark this answers is not in the document]
Comment (franz1981) [reply src/java.base/share/classes/java/lang/String.java]: separated question @cl4es, how do you obtained the 'Change` column with the `p` value? is in some jdk script I've missed?

**Comment by cl4es** [comment]
General comments:
- Does this suggest an optimization opportunity to `ArraysSupport::mismatch`, or is the relative overhead due the inherent costs of finding the exact index in the mismatch? 
- This only optimize the case where we're comparing the entirety of both `String`s. How common is this in practice? It would be good to check a mixed benchmark where some inputs will use this fast path and others won't to assert that this doesn't have an unexpected negative impact from adding a few tests.

**Comment by cl4es** [comment]
Filed: https://bugs.openjdk.org/browse/JDK-8321283

**Comment by franz1981** [comment]
> Does this suggest an optimization opportunity to ArraysSupport::mismatch, or is the relative overhead due the inherent costs of finding the exact index in the mismatch?

I suppose that's the latter; but I need to study the produced ASM for both, in case of small sized strings.
As said in https://github.com/openjdk/jdk/pull/16933#issuecomment-1838744137, the cost just fade away when cache accesses become the dominant cost

> This only optimize the case where we're comparing the entirety of both Strings. How common is this in practice?

I have the sole data point of HTTP endpoint routing which *can* have, in the hot path and common cases, exact matching too, see https://github.com/franz1981/vertx-web/blob/c6740a5991a15567521c26ac3a08091e8603f2f2/vertx-web/src/main/java/io/vertx/ext/web/impl/RouteState.java#L1251-L1262

> It would be good to check a mixed benchmark where some inputs will use this fast path and others won't to assert that this doesn't have an unexpected negative impact from adding a few tests.

You mean by having the different uses cases stressed in sequence with some probability (in order to induce some cache misses)? To be fair we should change the length too, and let the "batching" factor (ie the stride) of each different vectorized version to induce other branch-misses, instead of a "all or nothing" one, but sadly it means no longer having steady-state which will make more difficult to evaluate improvements. 
I'm +1 to this micro use case, but I would like to better understand how to write it right

File: src/java.base/share/classes/java/lang/String.java [hunk 3]
```diff
@@ -2177,10 +2177,14 @@ public boolean regionMatches(int toffset, String other, int ooffset, int len) {
...
            return true;
         }
+        byte coder = coder();
+        byte otherCoder = other.coder();
         byte[] tv = value;
         byte[] ov = other.value;
-        byte coder = coder();
-        if (coder == other.coder()) {
+        if (coder == otherCoder) {
+            if ((ooffset | toffset) == 0 && len == (tv.length >> coder) && ov.length == tv.length) {
```
[the remark this answers is not in the document]
Comment (franz1981) [reply src/java.base/share/classes/java/lang/String.java:2185]: It can be, depending the probability of happening for both and if/where they are inlined: conditions easy to predict are very cheap, but if not, can have some performance penalties. 
An `|` between register is super cheap, instead: the same tecnique has been used at https://github.com/openjdk/jdk/blob/a26f7c03c72e4efe6d3219d294294c554aebc631/src/hotspot/share/utilities/copy.cpp#L215 which is supposed to be another hot-path, probably for similar reasons.

**Comment by liach** [comment]
With a JBS issue, you need to change the PR title to `8321283: Reuse StringLatin1::equals in regionMatches` so that the JDK bot will recognize and link the issue.

File: src/java.base/share/classes/java/lang/String.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (franz1981) [reply src/java.base/share/classes/java/lang/String.java]: Gentle ping on the previous (although OT) question; did you have a OSS comparison tools which work on JMH results? If yes, is very nice!

**Comment by franz1981** [comment]
Keep it alive!

**Comment by liach** [comment]
@franz1981 Can you update the PR title to start with `8321283: ` so the jdk bot will recognize the JBS issue, mark this ready for review, and forward this PR to the OpenJDK mailing lists?

**Comment by RogerRiggs** [comment]
This PR has aged quite a bit. Please merge with master to get an up-to-date PR.
It might be cleaner to close this PR and start fresh.
Also, please reconfirm the performance numbers.
Thanks

**Comment by AlanBateman** [comment]
>  Please merge with master to get an up-to-date PR. It might be cleaner to close this PR and start fresh. Also, please reconfirm the performance numbers. Thanks

Yes, it looks this branch is nearly 2 years behind mainline so very important to sync up and refresh performance results.
